src/Hipace.cpp

```diff
@@ -507,39 +504,92 @@ Hipace::SolveOneSlice (int islice, const int islice_local, int step)
 
         if (m_explicit) {
-            ExplicitSolveOneSubSlice(lev, step, islice, islice_local);
+            // deposit jx, jy, chi and rhomjz for all plasmas
+            m_multi_plasma.DepositCurrent(m_fields, m_multi_laser, WhichSlice::This,
+                true, false, m_deposit_rho, true, true, m_3D_geom, lev);
+
+            // deposit jx_beam and jy_beam in the Next slice
+            m_multi_beam.DepositCurrentSlice(m_fields, m_3D_geom, lev, step, islice_local,
+                m_do_beam_jx_jy_deposition, false, false, WhichSlice::Next);
+
+            // deposit jz_beam and maybe rhomjz of the beam on This slice
+            m_multi_beam.DepositCurrentSlice(m_fields, m_3D_geom, lev, step, islice_local,
+                false, true, m_do_beam_jz_minus_rho, WhichSlice::This);
         } else {
-            PredictorCorrectorSolveOneSubSlice(lev, step, islice, islice_local);
+            // deposit jx jy jz (maybe chi) and rhomjz
+            m_multi_plasma.DepositCurrent(m_fields, m_multi_laser, WhichSlice::This,
+                true, true, m_deposit_rho, m_use_laser, true, m_3D_geom, lev);
+
+            // deposit jx jy jz and maybe rhomjz on This slice
+            m_multi_beam.DepositCurrentSlice(m_fields, m_3D_geom, lev, step, islice_local,
+                m_do_beam_jx_jy_deposition, true, m_do_beam_jz_minus_rho, WhichSlice::This);
         }
+        // add neutralizing background
+        m_fields.AddRhoIons(lev);
+
+        // deposit grid current into jz_beam
+        m_grid_current.DepositCurrentSlice(m_fields, m_3D_geom[lev], lev, islice);
+    }
+
+    // Psi ExmBy EypBx Ez Bz solve
+    m_fields.SolvePoissonPsiExmByEypBxEzBz(m_3D_geom, current_N_level);
+
+    // Bx By solve
+    if (m_explicit) {
+        for (int lev=0; lev<current_N_level; ++lev) {
+            // The algorithm used was derived in
+            // [Wang, T. et al. Phys. Rev. Accel. Beams 25, 104603 (2022)],
+            // it is implemented in the WAND-PIC quasistatic PIC code.
+
+            // Set Sx and Sy to beam contribution
+            InitializeSxSyWithBeam(lev);
 
+            // Deposit Sx and Sy for every plasma species
+            m_multi_plasma.ExplicitDeposition(m_fields, m_multi_laser, m_3D_geom, lev);
+
+            // Solves Bx, By using Sx, Sy and chi
+            ExplicitMGSolveBxBy(lev, WhichSlice::This);
+        }
+    } else {
+        // Solves Bx and By in the current slice and modifies the force terms of the plasma particles
+        PredictorCorrectorLoopToSolveBxBy(islice, islice_local, current_N_level, step);
+    }
+
+    if (m_multi_beam.isSalameNow(step, islice_local)) {
+        // Modify the beam particle weights on this slice to flatten Ez.
+        // As the beam current is modified, Bx and By are also recomputed.
+        SalameModule(this, m_salame_n_iter, m_salame_do_advance, m_salame_last_slice,
+                    m_salame_overloaded, current_N_level, step, islice, islice_local);
```

Comment:
So the SALAME is applied only to the finest level of the current slice, right? Just commenting, I don't see a practical case that could cause a problem here and would warrant an assert.